esp-hal/src/dma/mod.rs

```diff
@@ -902,141 +953,262 @@ impl DescriptorChain {
-            dw0.next = next;
+#[derive(Debug)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+struct DescriptorSet<'a> {
+    descriptors: &'a mut [DmaDescriptor],
+}
 
-            if last {
-                break;
+impl<'a> DescriptorSet<'a> {
+    /// Creates a new `DescriptorSet` from a slice of descriptors and associates
+    /// them with the given buffer.
+    fn new(descriptors: &'a mut [DmaDescriptor]) -> Result<Self, DmaBufError> {
+        if !is_slice_in_dram(descriptors) {
+            return Err(DmaBufError::UnsupportedMemoryRegion);
+        }
+
+        descriptors.fill(DmaDescriptor::EMPTY);
+
+        unsafe { Self::new_from_unchecked(descriptors) }
+    }
+
+    /// Creates a new `DescriptorSet` from a slice of descriptors and associates
+    /// them with the given buffer.
+    ///
+    /// # Safety
+    ///
+    /// The caller must ensure that the descriptors are in a supported memory
+    /// region.
+    unsafe fn new_from_unchecked(
+        descriptors: &'a mut [DmaDescriptor],
+    ) -> Result<Self, DmaBufError> {
+        Ok(Self { descriptors })
+    }
+
+    /// Consumes the `DescriptorSet` and returns the inner slice of descriptors.
+    fn into_inner(self) -> &'a mut [DmaDescriptor] {
+        self.descriptors
+    }
+
+    /// Returns a pointer to the first descriptor in the chain.
+    fn head(&mut self) -> *mut DmaDescriptor {
+        self.descriptors.as_mut_ptr()
+    }
+
+    /// Returns an iterator over the linked descriptors.
+    fn linked_iter(&self) -> impl Iterator<Item = &DmaDescriptor> {
+        let mut was_last = false;
+        self.descriptors.iter().take_while(move |d| {
+            if was_last {
+                false
+            } else {
+                was_last = d.next.is_null();
+                true
             }
+        })
+    }
 
-            processed += chunk_size;
-            descr += 1;
-        }
+    /// Returns an iterator over the linked descriptors.
+    fn linked_iter_mut(&mut self) -> impl Iterator<Item = &mut DmaDescriptor> {
+        let mut was_last = false;
+        self.descriptors.iter_mut().take_while(move |d| {
+            if was_last {
+                false
+            } else {
+                was_last = d.next.is_null();
+                true
+            }
+        })
+    }
 
-        Ok(())
+    /// Associate each descriptor with a chunk of the buffer.
+    ///
+    /// This function checks the alignment and location of the buffer.
+    ///
+    /// See `link_with_buffer_impl` for more details.
+    fn link_with_buffer(
+        &mut self,
+        buffer: &mut [u8],
+        chunk_size: usize,
+    ) -> Result<(), DmaBufError> {
+        Self::set_up_buffer_ptrs(buffer, self.descriptors, chunk_size, false)
     }
 
-    #[allow(clippy::not_unsafe_ptr_arg_deref)]
-    pub fn fill_for_tx(
+    /// Prepares descriptors for transferring `len` bytes of data.
+    ///
+    /// See `prepare_descriptors_impl` for more details.
```

Comment:
Who dis? :smile: 

If you use `[Self::prepare_descriptors_impl]` instead of `prepare_descriptors_impl`  the compiler would catch this.